lib/uplinkc/scope.go

```diff
@@ -32,3 +39,56 @@ func (parent *scope) child() scope {
 	ctx, cancel := context.WithCancel(parent.ctx)
 	return scope{ctx, cancel}
 }
+
+//export restrict_scope
+// restricts a given scope with the provided caveat and encryptionRestrictions
+func restrict_scope(scopeRef C.ScopeRef, caveatRef C.CaveatRef, restrictions **C.EncryptionRestriction, cerr **C.char) C.ScopeRef {
+	//Get apiKey from scope
+	scope, ok := universe.Get(scopeRef._handle).(*libuplink.Scope)
+	if !ok {
+		*cerr = C.CString("invalid scope")
+		return C.ScopeRef{}
+	}
+
+	//Get caveat from ref
+	caveat, ok := universe.Get(caveatRef._handle).(macaroon.Caveat)
+	if !ok {
+		*cerr = C.CString("invalid caveat")
+		return C.ScopeRef{}
+	}
+
+	//Restrict apiKey using caveat
+	//problem: caveat needs to know about encryptionsRestrictions before vv
+	apiKeyRestricted, err := scope.APIKey.Restrict(caveat)
+	if err != nil {
+		*cerr = C.CString("could not restrict apiKey")
+		return C.ScopeRef{}
+	}
+
+	//Convert EncryptionRestrictions to Go
+	restrictionsArray := (*[1 << 30 / unsafe.Sizeof(C.EncryptionRestriction{})]C.EncryptionRestriction)(unsafe.Pointer(restrictions))
+	restrictionsGo := make([]libuplink.EncryptionRestriction, 0)
+	for i := 0; i < len(restrictionsArray); i++ {
+		restrictionsGo = append(restrictionsGo, libuplink.EncryptionRestriction{
+			Bucket:     C.GoString(restrictionsArray[i].bucket),
+			PathPrefix: C.GoString(restrictionsArray[i].path_prefix),
+		})
+	}
+
+	//Create new EncryptionAccess with restrictions
+	//problem: it uses it's own caveat internally
+	_, encAccessRestricted, err := scope.EncryptionAccess.Restrict(apiKeyRestricted, restrictionsGo...) 
```

Comment:
you need to return `apikey` here and use it with constructing `libuplink.Scope`.

Reply:
Reworked the code